File: UINavigationExtensionDemo/Feature/Common/Views/FakeNavigationBar.m
//
//  FakeNavigationBar.m
//  UINavigationExtensionDemo
//
//  Created by Leo Lee on 2020/10/27.
//

#import "FakeNavigationBar.h"

#import "UIColor+RandomColor.h"
#import "UIDevice+Additions.h"

@interface FakeNavigationBar ()

@property (nonatomic, strong) UIButton *backButton;
@property (nonatomic, strong) UIButton *rightButton;
@property (nonatomic, strong) UILabel *titleLabel;

@end

@implementation FakeNavigationBar

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self addContentView];
    }
    return self;
}

#pragma mark - Private

- (void)addContentView {
    self.backgroundColor = [UIColor clearColor];
    self.titleLabel.alpha = 0.0;

    [self addSubview:self.backButton];
    [self addSubview:self.titleLabel];
    [self addSubview:self.rightButton];
    
    self.backButton.hidden = !UIDevice.isPhoneDevice;
    self.backButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.backButton.topAnchor constraintEqualToAnchor:self.topAnchor].active = YES;
    [self.backButton.bottomAnchor constraintEqualToAnchor:self.bottomAnchor].active = YES;
    [self.backButton.leftAnchor constraintEqualToAnchor:self.leftAnchor].active = YES;
    [self.backButton.widthAnchor constraintEqualToConstant:UIDevice.isPhoneDevice ? 44 : 0].active = YES;
    
    self.titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.titleLabel.topAnchor constraintEqualToAnchor:self.topAnchor].active = YES;
    [self.titleLabel.bottomAnchor constraintEqualToAnchor:self.bottomAnchor].active = YES;
    [self.titleLabel.leftAnchor constraintEqualToAnchor:self.backButton.rightAnchor].active = YES;
    [self.titleLabel.rightAnchor constraintEqualToAnchor:self.rightButton.leftAnchor].active = YES;
    
    self.rightButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.rightButton.topAnchor constraintEqualToAnchor:self.topAnchor].active = YES;
    [self.rightButton.bottomAnchor constraintEqualToAnchor:self.bottomAnchor].active = YES;
    [self.rightButton.rightAnchor constraintEqualToAnchor:self.rightAnchor].active = YES;
    [self.rightButton.widthAnchor constraintEqualToConstant:44].active = YES;
}

#pragma mark - Setter

- (void)setTitle:(NSString *)title {
    _title = title;
    self.titleLabel.text = title;
}

#pragma mark - Private

- (void)clickBackButton:(UIButton *)button {
    if (self.delegate && [self.delegate respondsToSelector:@selector(fakeNavigationBar:didClickNavigationItemWithItemType:)]) {
        [self.delegate fakeNavigationBar:self didClickNavigationItemWithItemType:FakeNavigationItemTypeBackButton];
    }
}

- (void)clickRightButton:(UIButton *)button {
    if (self.delegate && [self.delegate respondsToSelector:@selector(fakeNavigationBar:didClickNavigationItemWithItemType:)]) {
        [self.delegate fakeNavigationBar:self didClickNavigationItemWithItemType:FakeNavigationItemTypeAddButton];
    }
}

#pragma mark - Getter

- (UIButton *)backButton {
    if (!_backButton) {
        UIImage *image = [[UIImage imageNamed:@"NavigationBarBack"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        _backButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_backButton setImage:image forState:UIControlStateNormal];
        [_backButton addTarget:self action:@selector(clickBackButton:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _backButton;
}

- (UIButton *)rightButton {
    if (!_rightButton) {
        UIImage *image = [[UIImage imageNamed:@"NavigationBarAdd"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        _rightButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_rightButton setImage:image forState:UIControlStateNormal];
        [_rightButton addTarget:self action:@selector(clickRightButton:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _rightButton;
}

- (UILabel *)titleLabel {
    if (!_titleLabel) {
        _titleLabel = [[UILabel alloc] initWithFrame:CGRectZero];
        _titleLabel.font = [UIFont systemFontOfSize:18];
        _titleLabel.textAlignment = NSTextAlignmentCenter;
        _titleLabel.backgroundColor = [UIColor clearColor];
    }
    return _titleLabel;
}

#pragma mark - Public
- (void)updateNavigationBarAlpha:(CGFloat)alpha {
    UIColor *whiteColor = [UIColor colorWithRed:1.0 green:1.0 blue:1.0 alpha:1.0]; // 不要使用 [UIColor whiteColor]
    self.backButton.tintColor = [UIColor mixColor1:[UIColor customTitleColor] color2:whiteColor ratio:alpha];
    self.rightButton.tintColor = [UIColor mixColor1:[UIColor customTitleColor] color2:whiteColor ratio:alpha];
    self.titleLabel.alpha = alpha;
}

@end

File: Example/NXNavigationExtensionSwiftDemo/Common/Views/FakeNavigationBar.swift
//
//  FakeNavigationBar.swift
//  NXNavigationExtensionSwiftDemo
//
//  Created by Leo Lee on 2020/11/23.
//

import UIKit

protocol FakeNavigationBarDelegate: AnyObject {
    func fakeNavigationBar(_ navigationBar: FakeNavigationBar, didClickNavigationItemWithItemType itemType: FakeNavigationBar.ItemType)
}

class FakeNavigationBar: UIView {

    enum ItemType {
        case back, add
    }
    
    weak var delegate: FakeNavigationBarDelegate?
    var title: String? {
        didSet { titleLabel.text = title }
    }
    
    private var navigationBarAlpha = CGFloat(0.0)
    
    private lazy var backButton: UIButton = {
        let button = UIButton(type: .custom)
        button.setImage(UIImage(named: "NavigationBarBack")?.withRenderingMode(.alwaysTemplate), for: .normal)
        button.addTarget(self, action: #selector(clickBackButton(_:)), for: .touchUpInside)
        button.translatesAutoresizingMaskIntoConstraints = false
        return button
    }()
    
    private lazy var addButton: UIButton = {
        let button = UIButton(type: .custom)
        button.setImage(UIImage(named: "NavigationBarAdd")?.withRenderingMode(.alwaysTemplate), for: .normal)
        button.addTarget(self, action: #selector(clickAddButton(_:)), for: .touchUpInside)
        button.translatesAutoresizingMaskIntoConstraints = false
        return button
    }()
    
    private lazy var titleLabel: UILabel = {
        let label = UILabel(frame: .zero)
        label.font = .systemFont(ofSize: 18)
        label.textAlignment = .center
        label.backgroundColor = .clear
        label.translatesAutoresizingMaskIntoConstraints = false
        return label
    }()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupViewContent()
        addViewContent()
        addViewConstraint()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        updateNavigationBar(alpha: navigationBarAlpha)
    }
    
}

extension FakeNavigationBar {
    
    private func setupViewContent() {
        backgroundColor = .clear
        titleLabel.alpha = 0.0
        backButton.isHidden = !UIDevice.isPhone
    }
    
    private func addViewContent() {
        addSubview(backButton)
        addSubview(titleLabel)
        addSubview(addButton)
    }
    
    private func addViewConstraint() {
        backButton.topAnchor.constraint(equalTo: topAnchor).isActive = true
        backButton.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
        backButton.leftAnchor.constraint(equalTo: leftAnchor).isActive = true
        backButton.widthAnchor.constraint(equalToConstant: UIDevice.isPhone ? 44.0 : 0.0).isActive = true
        
        titleLabel.topAnchor.constraint(equalTo: topAnchor).isActive = true
        titleLabel.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
        titleLabel.leftAnchor.constraint(equalTo: backButton.rightAnchor).isActive = true
        titleLabel.rightAnchor.constraint(equalTo: addButton.leftAnchor).isActive = true
     
        addButton.topAnchor.constraint(equalTo: topAnchor).isActive = true
        addButton.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
        addButton.rightAnchor.constraint(equalTo: rightAnchor).isActive = true
        addButton.widthAnchor.constraint(equalToConstant: 44.0).isActive = true
    }
    
}

@objc
extension FakeNavigationBar {
    
    private func clickBackButton(_ button: UIButton) {
        delegate?.fakeNavigationBar(self, didClickNavigationItemWithItemType: .back)
    }
    
    private func clickAddButton(_ button: UIButton) {
        delegate?.fakeNavigationBar(self, didClickNavigationItemWithItemType: .add)
    }
    
}

extension FakeNavigationBar {
    
    func updateNavigationBar(alpha: CGFloat) {
        navigationBarAlpha = alpha
        
        let white = UIColor(red: 1.0, green: 1.0, blue: 1.0, alpha: 1.0 ) // 不要使用 .white
        let change = UIColor.customDynamic {
            return UIColor(red: 0.0, green: 0.0, blue: 0.0, alpha: 1.0) // .black
        } darkMode: {
            return UIColor(red: 1.0, green: 0.0, blue: 0.0, alpha: 1.0) // .red
        }
        
        backButton.tintColor = UIColor.mix(color1: change, color2: white, ratio: alpha)
        addButton.tintColor = UIColor.mix(color1: change, color2: white, ratio: alpha)
        titleLabel.alpha = alpha
    }
    
}
